### PR TITLE
341: Fix Validation Rule Engine concurrency issues

### PR DESCRIPTION
**Pull request by basiliskus**
# Fix Validation Rule Engine concurrency issues

- Fixed concurrency issues when loading rules
- Fixed validation rule fhirpath for "Message has required receiver id". It was previously matching receiver namespace as well

## Issue

#341 

## Checklist

- [x] I have added tests to cover my changes
- [x] I have added logging where useful (with appropriate log level)
- [x] I have added JavaDocs where required
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngine.java [hunk 1]
```diff
@@ -28,7 +28,7 @@ public void unloadRules() {
         rules.clear();
     }
 
-    public void ensureRulesLoaded() {
+    public synchronized void ensureRulesLoaded() {
```
Comment (halprin) [inline etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngine.java:31]: This is in a performance sensitive flow during API flows, and synchronizing here will slow down concurrent calls to our API.

We can prevent this bottleneck by doing "double-checked locking".

```java
public void ensureRulesLoaded() {
    if (rules.isEmpty()) {
        synchronized(this) {
            if (rules.isEmpty()) {
                loadRules();
            }
        }
    }
}
```

This also allows us to remove `synchronized` from the `loadRules` method below.

**Review by halprin — CHANGES_REQUESTED** [review]

File: etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngine.java (hunk 1 above)
Comment (luis-pabon-tf) [reply etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngine.java:31]: That's a really good catch! Updating...

File: etor/src/test/groovy/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngineTest.groovy [hunk 2]
```diff
@@ -39,6 +39,27 @@ class RuleEngineTest extends Specification {
         1 * mockRuleLoader.loadRules(_ as String) >> [Mock(Rule)]
     }
 
+    def "ensureRulesLoaded loads rules only once on multiple threads"() {
+        given:
+        def threadsNum = 2
```
Comment (halprin) [inline etor/src/test/groovy/gov/hhs/cdc/trustedintermediary/etor/ruleengine/RuleEngineTest.groovy:44]: Let's up this to 10 so we have more threads in contention for the lock.  It's easy for 2 to not execute the method at the same time.

```suggestion
        def threadsNum = 10
```

**Review by halprin — APPROVED** [review]
Looks great!

**Pull request merged**